jazkarta/shop/browser/controlpanel.py

```diff
@@ -67,43 +74,252 @@ def _fetch_orders(part, key=()):
+    start_index = 0
+    orders_exist = False
 
     def update(self):
-        orders = list(_fetch_orders(storage.get_storage()))
+        orders = list(_fetch_orders(storage.get_storage(), (), False))
         orders.sort(key=lambda o: o.get('date_sort', ''), reverse=True)
         start = int(self.request.get('b_start', 0))
+
+        if len(orders) > 0:
+            self.orders_exist = True
+
+            self.most_recent_order_date = orders[0]['date']
+            self.first_order_date = orders[len(orders)-1]['date']
+
+            # default in case date selection integrity check fails
+            # this could happen if end date < start date
+            self.end_index = 0 
+            self.start_index = len(orders)-1
+
+            selected_start = self.to_datetime(self.startDateString(),'%Y-%m-%d')
+            selected_end = self.to_datetime(self.endDateString(),'%Y-%m-%d')
+
+            if self.check_date_integrity():
+                # generate list of dates in selected date range
+                count = selected_start
+                date_range = []
+                while count <= selected_end:        
+                    date_range.append(count)
+                    count += datetime.timedelta(days=1)
+
+                # find orders indexes that are in selected range
+                indexes = [ orders.index(x) for x in orders if \
+                    self.to_datetime(x['date'], \
+                    '%Y-%m-%d %I:%M %p').replace(hour=0, minute=0) in date_range]
+                if len(indexes) > 0:
+                    self.end_index = indexes[0] # newest 
+                    self.start_index = indexes[len(indexes)-1] # oldest
+                    # trim orders based on selected start and end 
+                    orders = orders[self.end_index:self.start_index]
+                else:
+                    # no orders in selected date range
+                    orders = []
+
         self.batch = Batch(orders, size=50, start=start)
         super(OrderControlPanelView, self).update()
 
 
+class ExportShopOrders(BrowserView, DateMixin):
+    """ Export selected date range of shop orders into a CSV file, do nothing
+        if no orders exist.
+    """
+
+    def __call__(self):
+        csv_content = None
+        # get shop order entries
+        orders = list(_fetch_orders(storage.get_storage(), (), True))
+        orders.sort(key=lambda o: o.get('date_sort', ''), reverse=True)
+        orders_csv = StringIO()
+
+        # get indexes of selected orders dates if supplied
+        first_order = self.request.get('first_order')
+        last_order = self.request.get('last_order')
+        if first_order and last_order:
+            try:
+                if int(last_order) <= len(orders)-1 and int(first_order) >= 0:
+                    # trim selection
+                    orders = orders[int(first_order):int(last_order)]
+            except:
+                pass
+
+        if orders is not None and len(orders) > 0:
```

Comment:
You can just write `if orders:` (empty lists are false-y)

Reply:
:+1: